cmd/apiserver-watcher/azure_handler.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/golang/glog"
+)
+
+// Prevent hairpin traffic when the apiserver is up
+
+// As per the Azure documentation (https://docs.microsoft.com/en-us/azure/load-balancer/concepts//limitations),
+// if a backend is load-balanced to itself, then the traffic will be dropped.
+//
+// This is because the L3LB does DNAT, so while the outgoing packet has a destination
+// IP of the VIP, the incoming load-balanced packet has a destination IP of the
+// host. That means that it "sees" a syn with the source and destination
+// IPs of itself, and duly replies wit a syn-ack back to itself. However, the client
+// socket expects a syn-ack with a source IP of the VIP, so it drops the packet.
+//
+// The solution is to redirect traffic destined to the lb vip back to ourselves.
+
+type azureHandler struct {
+	vip       string
+	ip4tables *iptables.IPTables
+	ip6tables *iptables.IPTables
+	stopCh    chan struct{}
+}
+
+var _ handler = &azureHandler{}
+
+func newAzureHandler(vip string) (handler, error) {
+	ipt4, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(5),
+	)
+	if err != nil {
+		return nil, err
+	}
+	ipt6, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv6),
+		iptables.Timeout(iptablesTimeout),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &azureHandler{
+		vip:       vip,
+		ip4tables: ipt4,
+		ip6tables: ipt6,
+		stopCh:    make(chan struct{}),
+	}, nil
+}
+
+// onFailure: either stop the routes service, or write downfile
+func (az *azureHandler) onFailure() error {
+	select {
+	case az.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+	return az.deleteRules()
+}
+
+// onSuccess: either start routes service, or remove down file
+func (az *azureHandler) onSuccess() error {
+	if err := az.syncRulesOnce(); err != nil {
+		return err
+	}
+	go az.syncRulesUntil()
+	return nil
+
+}
+
+// https://github.com/openshift/installer/blob/master/data/data/azure/vnet/internal-lb.tf
+// interval_in_seconds = 5
+// number_of_probes    = 2
+func (az *azureHandler) successThreshold() int {
+	return 1
+}
+
+func (az *azureHandler) failureThreshold() int {
+	// LB = 6 seconds, plus 10 seconds for propagation
+	return 8
+}
```

Comment:
I took the existing values but it will be good if somebody can review them, I think that azure and gcp lb using different parameters but the previous code was using the same for both

Reply:
/cc @sttts

Reply:
no idea on this did anyone double check?

Reply:
this doesn't change the timers, so it will keep working as before